src/soter/openssl/soter_ec_key_3.c

```diff
@@ -0,0 +1,391 @@
+
+err:
+
+    EVP_PKEY_CTX_free(ctx);
+
+    return res;
+}
+
+soter_status_t soter_ec_priv_key_to_engine_specific(const soter_container_hdr_t* key,
+                                                    size_t key_length,
+                                                    soter_engine_specific_ec_key_t** engine_key)
+{
+    const char* curve_str;
+    int curve;
+    OSSL_PARAM* params = NULL;
+    OSSL_PARAM_BLD* bld = NULL;
+    EVP_PKEY_CTX* ctx = NULL;
+    BIGNUM* d = NULL;
+    soter_status_t res;
+
+    if (key_length != be32toh(key->size)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    /* Validate tag */
+    if (memcmp(key->tag, EC_PRIV_KEY_PREF, strlen(EC_PRIV_KEY_PREF)) != 0) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (SOTER_SUCCESS != soter_verify_container_checksum(key)) {
+        return SOTER_DATA_CORRUPT;
+    }
+
+    switch (key->tag[3]) {
+    case EC_SIZE_TAG_256:
+        curve_str = SN_X9_62_prime256v1;
+        curve = NID_X9_62_prime256v1;
+        break;
+    case EC_SIZE_TAG_384:
+        curve_str = SN_secp384r1;
+        curve = NID_secp384r1;
+        break;
+    case EC_SIZE_TAG_521:
+        curve_str = SN_secp521r1;
+        curve = NID_secp521r1;
+        break;
+    default:
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (key_length < ec_priv_key_size(curve)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    bld = OSSL_PARAM_BLD_new();
+    if (bld == NULL) {
+        res = SOTER_NO_MEMORY;
+        goto err;
+    }
+
+    if (!OSSL_PARAM_BLD_push_utf8_string(bld, OSSL_PKEY_PARAM_GROUP_NAME, curve_str, strlen(curve_str))) {
+        res = SOTER_FAIL;
+        goto err;
+    }
+
+    d = BN_bin2bn((const unsigned char*)(key + 1), (int)(key_length - sizeof(soter_container_hdr_t)), NULL);
+    if (d == NULL) {
+        res = SOTER_NO_MEMORY;
+        goto err;
+    }
+
+    // Unfortunately, using on stack `OSSL_PARAM params[3]`
+    // like in soter_ec_pub_key_to_engine_specific() and setting
+    // params[1] = OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_PRIV_KEY,
+    //                                     (unsigned char*)(key + 1) + 1,
+    //                                     (int)(key_length - sizeof(soter_container_hdr_t) - 1)))
+    // won't work properly. This is because private key number (32 bytes, not counting leading zero)
+    // is stored in one byte order and EVP_PKEY_fromdata() will read it in different byte order. As
+    // a result, `priv` property of generated `pkey` will have different value, with reversed bytes
+    // order. One working solution was to do like this, deserialize value into `BIGINT*`, then use
+    // param builder (OSSL_PARAM_BLD*) and OSSL_PARAM_BLD_push_BN(). Seems to have a slight overhead,
+    // but uses APIs that were designed for such actions.
+    // Another possible and working solution is to simply swap bytes into temporary buffer and use
+    // it instead of provided one, then we can avoid using EVP_PKEY_BLD* at all.
```

Comment:
Yeah, their solution seems faster than mine. But that function accepts `BIGNUM*` so it is forced to serialize, while here in Themis we receive buffer so could avoid using `BIGINT` at all and just copy bytes in reversed order into temporary buffer on stack. It's just the fact that this solution is a bit... quirky, maybe not what OpenSSL devs expected. Guess we could consider it if benchmarks show significant difference. Also, on big-endian architectures (if Themis supports one) it could be a no-op